---
title: "RNAseq_report"
author: "Dr. S. Virdi"
date: "..."
output:  
  html_document:
    toc: true
---
<style type="text/css">
.main-container {
  max-width: 1800px;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicFeatures)
library(tximport)
library(DESeq2)
library(ggplot2)
library(clusterProfiler)
library(plyr)
library(topGO)
library(genefilter)
library(fdrtool)
library(ggrepel)
library(ggpubr)
library(AnnotationDbi)
library(openxlsx)
library(DT)
library(htmlwidgets)
library(pheatmap)
library(readr)
library(org.Hs.eg.db)
library(vsn)


```
## Analysis  {.tabset}

### Initialization
Importing ENSEMBL annotations and salmon transcript counts
```{r, echo=FALSE}

HS<-"hsapiens_gene_ensembl"
MM<-"mmusculus_gene_ensembl"
DM<-"dmelanogaster_gene_ensembl"
MA <-"mauratus_gene_ensembl"

#/mnt/HPI1/Ref/M_auratus/Mesocricetus_auratus.MesAur1.0.101.gtf
#/mnt/HPI1/Ref/h_sapiens/gencode.v33.annotation.gtf
#ensembl = useMart(biomart="ensembl", dataset=MA, host="uswest.ensembl.org")
txdb <- makeTxDbFromGFF("/Volumes/data_external_1tb/Ref/h_sapiens/gencode.v43.annotation.gtf",
                 format=c("auto", "gff3", "gtf"))

k <- keys(txdb, keytype = "TXNAME")
tx2gene <- AnnotationDbi::select(txdb, k, "GENEID", "TXNAME")

```

Making DEseq2 objects
```{r, echo=FALSE, message = FALSE, warning = FALSE}
control <- ct
stim <- sm
grp  <- c(rep("CT",length(control)),rep("TR",length(stim)))
cData  <- data.frame(condition= as.factor(grp))
design.matrix  <- model.matrix(~ condition, data  = cData)
txi.salmon <- tximport(c(control,stim), type = "salmon", tx2gene = tx2gene, ignoreAfterBar = TRUE)
rownames(cData) <- colnames(txi.salmon$counts)
dds_salmon <- DESeqDataSetFromTximport(txi.salmon,
                                colData = cData,
                                design = ~ condition)
colData(dds_salmon)$condition <- factor(colData(dds_salmon)$condition, levels = c("CT", "TR"))
```

### QC
```{r, echo=FALSE, warning = FALSE}
GeneCounts <- counts(dds_salmon)
idx.nz <- apply(GeneCounts, 1, function(x) { all(x > 0)})
cat("total number of genes with non-zero counts in all sample is", sum(idx.nz))
```

### Normalization and PCA 
```{r, echo=FALSE, message = FALSE, warning = FALSE}
dds <- estimateSizeFactors(dds_salmon, type="iterate")
#qs <- apply(counts(dds), 2, quantile, .9)
#sf <- qs / exp(mean(log(qs)))
#sizeFactors(dds) <- sf

SF<-data.frame(t(sizeFactors(dds)))
cat("Size factors")
SF
#write.xlsx(SF, file = "size_factors.xlsx")


normalised.table.dds <- (counts(dds,normalized=T))
rownames(normalised.table.dds) <- gsub('\\..+$', '', rownames(normalised.table.dds))
                                                                                                     ####################
                                                                                                     ##                ##
                                                                                                     ## DIR-CHANGE     ##
                                                                                                     ##                ##
                                                                                                     ####################

dr="..."
write.table(normalised.table.dds, file = paste0(dr,"normalizedCounts.table"), sep = "\t")
#write.xlsx2(normalised.table.dds, file = paste0(dr,"normalizedCounts.xlsx"))

vsn <- varianceStabilizingTransformation(dds, blind=TRUE)
PCA <- prcomp(t(assay(vsn)), scale = FALSE) 
percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)


PCA_data = data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2], 
                    PC3 = PCA$x[,3], PC4 = PCA$x[,4],
                    sample = rownames(colData(vsn)),
                    condition = colData(vsn)$condition)



#png("PCA.png",units="in", width=14, height=8, res=600)
ggplot(PCA_data, aes(x=PC1,y=PC2, label = rownames(PCA_data), color=condition)) +  
    geom_point(size=3.1, stroke = 2) +   geom_text_repel(size=3.15,stat = "identity") +
    theme(axis.text.x  = element_text(size = 3.5,margin = margin(r = 0))) +
    theme(axis.text.y = element_text(size = 3.5,margin = margin(r = 0))) +
    xlab(paste0("PC1, VarExp:", round(percentVar[1],4))) + 
    ylab(paste0("PC2, VarExp:", round(percentVar[2],4))) + 
    theme(axis.title.y = element_text(size = 10.5))+
    theme(axis.title.x = element_text(size = 10.5))+
    theme(panel.background = element_rect(fill="white"),
          axis.line = element_line(colour = "black", size = 0.0),
          panel.grid.minor = element_line(color="grey",size = 0.1),
          panel.grid.major = element_line(color="grey",size = 0.1),
          axis.ticks = element_blank(),
          panel.border = element_rect(colour = "black", fill=NA, size=0.2)) + 
    scale_color_manual(values=c("#e95462","black"))  + scale_shape_manual(values = c(19,1)) +
    theme(legend.position="none") 
#dev.off()

```

### DE Analysis
```{r,echo=FALSE, message = FALSE, warning = FALSE}
dds <- estimateDispersions(dds)
#plotDispEsts(dds)
#keep <- rowSums(counts(dds)) > 1
#dds <- dds[keep,]

dds<-  nbinomWaldTest(dds)
DESeq2Res <- results(dds, pAdjustMethod = "BH")
table(DESeq2Res$padj < 0.1)
#table(DESeq2Res$padj < 0.05)
#QC DE
#hist(DESeq2Res$pvalue, col = "lavender", main = "WT vs Deletion", xlab = "p-values")

DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$padj), ]

DESeq2Res <- DESeq2Res[ !is.na(DESeq2Res$pvalue), ]

DESeq2Res <- DESeq2Res[, -which(names(DESeq2Res) == "padj")]
suppressMessages(FDR.DESeq2Res <- fdrtool(DESeq2Res$stat, statistic= "normal", plot = FALSE))
#FDR.DESeq2Res$param[1, "sd"]
DESeq2Res[,"padj"]  <- p.adjust(FDR.DESeq2Res$pval, method = "BH")
table(DESeq2Res$padj < 0.1)
#QC
#hist(FDR.DESeq2Res$pval, col = "royalblue4", main = "CT vs TR, correct null model", xlab = "CORRECTED p-values")

#table(DESeq2Res[,"padj"] < 0.1)

rownames(DESeq2Res) <- gsub('\\..+$', '', rownames(DESeq2Res))
gene_ensembl_id<-data.frame(Ensembl_id=rownames(DESeq2Res))

#genes_names<-merge(gene_ensembl_id, getBM(filters=c("ensembl_gene_id"), 
#                                          attributes=c("ensembl_gene_id","mgi_symbol"), 
#                                          values=res$Ensembl_id, mart=ensembl), by.x="Ensembl_id",by.y="ensembl_gene_id")

## OR ###

genes_names<-merge(gene_ensembl_id, AnnotationDbi::select(org.Hs.eg.db, keys=as.vector(gene_ensembl_id$Ensembl_id),
            columns=c("SYMBOL","ENTREZID"), keytype="ENSEMBL"), by.x="Ensembl_id",by.y="ENSEMBL")


dup_row<-which(duplicated(genes_names$Ensembl_id))
if (length(dup_row) != 0) {
  genes_names<-genes_names[-which(duplicated(genes_names$Ensembl_id)),]
}



sigGenes<-data.frame(subset(DESeq2Res, padj < 0.1))

sigGenes<-cbind(sigGenes, ensembl_id=rownames(sigGenes))

#sig_genes_names<- merge(sigGenes, getBM(filters=c("ensembl_gene_id"), 
#                                        attributes=c("ensembl_gene_id","external_gene_name","entrezgene"), 
#                                        values=sigGenes$ensembl_id, mart=ensembl), by.x="ensembl_id",by.y="ensembl_gene_id")

sig_genes_names<- merge(sigGenes, AnnotationDbi::select(org.Hs.eg.db, keys=as.vector(sigGenes$ensembl_id),
                 columns=c("SYMBOL","ENTREZID"), keytype="ENSEMBL"), by.x="ensembl_id",by.y="ENSEMBL")


write.xlsx(merge(sig_genes_names,normalised.table.dds, by.x="ensembl_id", by.y="row.names"), file = paste0(dr,"DE_norm.xlsx"))



                                                                                           ###################################
                                                                                           ##            PLOT               ##
                                                                                           ##                               ##
                                                                                           ###################################

plot_genes <- gene_ensembl_id
rownames(plot_genes) <- plot_genes$Ensembl_id
plot_genes[sig_genes_names$ensembl_id,] <- sig_genes_names$SYMBOL
#png("MA_plot.png", units="in", width=8, height=6, res=800)
suppressWarnings(data.frame(DESeq2Res) %>%
  dplyr::mutate(out_of_bounds = abs(log2FoldChange) > 5,
    log2FoldChange = ifelse(out_of_bounds, 5*sign(log2FoldChange), log2FoldChange)) %>% 
  ggmaplot(main = expression("FDR < 0.1)"),
           fdr = 0.1, fc = 1^2, size = 0.4,
           palette = c("#B31B21", "#1465AC", "darkgray"),
           genenames = as.vector(plot_genes$Ensembl_id),
           legend = "top", top = 20,
           font.label = c("bold", 11), label.rectangle = TRUE,
            font.legend = "bold",
             font.main = "bold",
              ggtheme = ggplot2::theme_minimal()))

#genenames = as.vector(genes_names$SYMBOL)
#as.vector(rownames(DESeq2Res_plot))
#dev.off()

##Volcano

DESeq2Res_plot$diffexpressed <- "NO"
DESeq2Res_plot$diffexpressed[DESeq2Res_plot$log2FoldChange > 0.5 & DESeq2Res_plot$padj < 0.1] <- "UP"
DESeq2Res_plot$diffexpressed[DESeq2Res_plot$log2FoldChange < -0.5 & DESeq2Res_plot$padj < 0.1] <- "DOWN"


#tiff("Volcano.tiff", units="in", width=6.6, height=5, res=150)
ggplot(DESeq2Res_plot, aes(x=log2FoldChange, y=-log10(padj), col=diffexpressed)) +
  geom_point(size=0.2) + scale_color_manual(values=c("blue", "black", "red")) +
  theme(panel.background = element_rect(fill="white"),
        axis.line = element_line(colour = "black", size = 0.0),
        panel.grid.minor = element_line(color="grey",size = 0.1),
        panel.grid.major = element_line(color="grey",size = 0.1),
        axis.ticks = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, size=0.2), aspect.ratio = 1) +
  labs(title = "FDR = 0.1")
#dev.off()






##Heatmap


hm_vsn <- as.data.frame(assay(vsn))
x <- rownames(hm_vsn)
#x[c( 62277,62278 ,62283 ,62288)] <- c("K4_1","K4_2", "K8_1", "ORF17_5")
rownames(hm_vsn) <- x
rownames(hm_vsn) <- gsub('\\..+$', '',rownames(hm_vsn))

pheatmap(hm_vsn[intersect(sig_genes_names$ensembl_id, rownames(hm_vsn)),],scale = "row",clustering_distance_rows="correlation", show_rownames = FALSE, cluster_cols = FALSE,treeheight_row = 0)


  ##Mus_musculus
  ##Drosophila_melanogaster
  ##Homo_sapiens

DT_sig_genes_names<-sig_genes_names
DT_sig_genes_names$ensembl_id <- paste0('<a href="',"http://www.ensembl.org/","Homo_sapiens","/Gene/Summary?db=core;g=",sig_genes_names$ensembl_id,'">',sig_genes_names$ensembl_id,"</a>")

datatable(DT_sig_genes_names,extensions = 'Buttons', filter = 'top', options = list(pageLength = 100, scrollX = TRUE, scrollY = "600px", autoWidth = TRUE, dom = 'Bfrtip',
    buttons = 
      list('copy', 'print', list(
        extend = 'collection',
        buttons = c('csv', 'excel'),
        text = 'Download'
      ))), escape = FALSE)
```

### GO_tables
```{r,echo=FALSE, message = FALSE, warning = FALSE}

                                                                                                ##################
                                                                                                ##              ##
                                                                                                ## DIR CHANGE ####
                                                                                                ##              ##
                                                                                                ##################

dr="..."

BaseMeanTotal <- as.matrix(DESeq2Res[, "baseMean", drop = FALSE])
signames <- match(sig_genes_names$ensembl_id, rownames(BaseMeanTotal))
background <- c()

for(i in signames){
  ii <- genefilter::genefinder(BaseMeanTotal, i, 10, method = "manhattan")[[1]]$indices
  background <- c(background, ii)
}

background <- unique(background)
background <- rownames(BaseMeanTotal)[background]

background <- setdiff(background,  sig_genes_names$ensembl_id)
#length(backG)
#QC
#png("Density_plot.png", units="in", width=8, height=6, res=800)
## multidensity(list(all= log2(DESeq2Res[,"baseMean"]) ,  foreground =log2(DESeq2Res[sig_genes_names$ensembl_id, "baseMean"]),  background =log2(DESeq2Res[backG, "baseMean"])),  xlab="log2 mean normalized counts", main = "Matching for enrichment analysis")
#dev.off()

gos = c( "MF", "BP", "CC" )

gene_id = rownames(BaseMeanTotal)
universe = gene_id %in% c(as.character(sig_genes_names$ensembl_id), background) 
in_selection =  gene_id %in%  c(as.character(sig_genes_names$ensembl_id))

allgenes <- factor( as.integer(in_selection[universe]))
names(allgenes) <- gene_id[universe]

tab = as.list(gos)
names(tab) = gos



for(i in 1:3){
  
  ## prepare data
  go_data <- new("topGOdata", ontology=gos[i], allGenes = allgenes, nodeSize=5,
              annot=annFUN.org, mapping="org.Hs.eg.db", ID = "ensembl" )
  
  ## run tests
  elim <- runTest(go_data, algorithm = "elim", statistic = "Fisher" )
  classic <- runTest(go_data, algorithm = "classic", statistic = "Fisher" )
  
  ## look at results
  tab[[i]] <- GenTable( go_data, Fisher.elim = elim, 
                        Fisher.classic = classic,
                        orderBy = "Fisher.classic", topNodes = 200)
  
   #, topNodes = 100
}

subDir<-"/GO_html"
dir.create(file.path(paste0(dr, subDir)), showWarnings = FALSE)
setwd(file.path(paste0(dr, subDir)))


go_data <- new("topGOdata", ontology=gos[2], allGenes = allgenes, nodeSize=5,
           annot=annFUN.org, mapping="org.Hs.eg.db", ID = "ensembl" )

topGO_BP <- tab$BP
#To find genes related to each GO term
AnnotatedGenes_BP <- lapply(topGO_BP$GO.ID, function(x) as.character(unlist(genesInTerm(object = go_data, whichGO = x)))) # list containig genes annotated to significant GO terms
#Put lined gene list in form of html file and save it and link it to each GO term
topGO_BP$GO.ID<-gsub(":","_",topGO_BP$GO.ID)


for (go in 1:length(topGO_BP$GO.ID)){
  ##Mus_musculus
  ##Drosophila_melanogaster
  ##Homo_sapiens
  
  go_gene_list <- sig_genes_names[sig_genes_names$ensembl_id %in% intersect(sig_genes_names$ensembl_id,AnnotatedGenes_BP[[go]]),]
  go_gene_list$ensembl_id <-paste0('<a href="',"http://www.ensembl.org/","Homo_sapiens","/Gene/Summary?db=core;g=",go_gene_list$ensembl_id,'">',go_gene_list$ensembl_id,"</a>")
  go_gene_list<-datatable(go_gene_list, escape = FALSE)
  htmlwidgets::saveWidget(go_gene_list, paste0(topGO_BP$GO.ID[[go]],".html"))
}


topGO_BP$GO.ID <- paste0('<a href="',"./GO_html/",topGO_BP$GO.ID,".html",'">', topGO_BP$GO.ID,"</a>")

datatable(topGO_BP,filter = 'top', options = list(pageLength = 100, scrollX = TRUE, scrollY = "600px", autoWidth = TRUE), escape = FALSE)


```

### GO_plots
```{r,echo=FALSE, message = FALSE, warning = FALSE}


allgenes_names_ensembl<-data.frame(ensembl_id=gene_id[universe])

allgenes_names <- merge(allgenes_names_ensembl, AnnotationDbi::select(org.Hs.eg.db,              
                  keys=as.vector(allgenes_names_ensembl$ensembl_id),
                  columns=c("SYMBOL","ENTREZID"), keytype="ENSEMBL"),
                  by.x="ensembl_id",by.y="ENSEMBL")



CP_res <-enrichGO(gene = as.vector(sig_genes_names$ENTREZID),
         universe = as.character(allgenes_names$ENTREZID),
         OrgDb         = "org.Hs.eg.db",
         ont           = "BP",
         pAdjustMethod = "BH",
         pvalueCutoff  = 0.05,
         qvalueCutoff  = 0.2,
         readable      = TRUE)

clusterProfiler::dotplot(CP_res)
tiff(paste0(dr,"GO_clusterprofiler.tiff"), units="cm", width=20, height=24, res=80)
clusterProfiler::dotplot(CP_res,showCategory= 30)
dev.off()


```

### Report information
```{r,echo=FALSE, message = FALSE, warning = FALSE}
sessioninfo::session_info(pkgs = c("attached")[1], dependencies = "NO")
```

